### PR TITLE
add backup related crd type and initialize controller skeleton

### PR DESCRIPTION
**Pull request by onlymellb**
### What problem does this PR solve? 
Refactoring backup and recovery logic

### What is changed and how does it work?
This pr just add backup related CRD type and initialize controller skeleton, and the following prs will gradually realize the whole function.

### Check List 

Tests 

 - Unit test
 - E2E test
 - Stability test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has Go code change

Side effects

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:
 
 ```release-note
add backup related CRD type and initialize controller skeleton
 ```

### CODE REVIEW

File: pkg/apis/pingcap.com/v1alpha1/types.go [hunk 1]
```diff
@@ -258,3 +258,215 @@ type TiKVFailureStore struct {
...
+	BackupRunning BackupConditionType = "Running"
+	// BackupComplete means the Backup has successfully executed and the
+	// resulting artifact has been stored in backend storage.
+	BackupComplete BackupConditionType = "Complete"
+	// BackupFailed means the Backup has failed.
+	BackupFailed BackupConditionType = "Failed"
+)
+
+// BackupCondition describes the observed state of a Backup at a certain point.
+type BackupCondition struct {
```
Comment (LinuxGit) [inline pkg/apis/pingcap.com/v1alpha1/types.go]: no json struct field tags

File: pkg/apis/pingcap.com/v1alpha1/types.go [hunk 2]
```diff
@@ -258,3 +258,215 @@ type TiKVFailureStore struct {
...
+	// CommitTs is the snapshot time point of tidb cluster.
+	CommitTs   string            `json:"backupSize"`
+	Conditions []BackupCondition `json:"conditions"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// BackupSchedule is a backup schedule of tidb cluster.
+type BackupSchedule struct {
```
Comment (weekface) [inline pkg/apis/pingcap.com/v1alpha1/types.go]: This should be `ScheduledBackup`

File: pkg/apis/pingcap.com/v1alpha1/types.go [hunk 3]
```diff
@@ -258,3 +258,215 @@ type TiKVFailureStore struct {
...
+	// BackupPath is the location of the backup.
+	BackupPath string `json:"backupPath"`
+	// TimeStarted is the time at which the backup was started.
+	TimeStarted metav1.Time `json:"timeStarted"`
+	// TimeCompleted is the time at which the backup completed.
+	TimeCompleted metav1.Time `json:"timeCompleted"`
+	// BackupSize is the data size of the backup.
+	BackupSize int64 `json:"backupSize"`
+	// CommitTs is the snapshot time point of tidb cluster.
+	CommitTs   string            `json:"backupSize"`
```
Comment (LinuxGit) [inline pkg/apis/pingcap.com/v1alpha1/types.go]: s/backupSize/commitTs

File: pkg/apis/pingcap.com/v1alpha1/types.go [hunk 4]
```diff
@@ -258,3 +258,215 @@ type TiKVFailureStore struct {
...
+	RestoreRunning RestoreConditionType = "Running"
+	// RestoreComplete means the Restore has successfully executed and the
+	// backup data has been loaded into tidb cluster.
+	RestoreComplete RestoreConditionType = "Complete"
+	// RestoreFailed means the Restore has failed.
+	RestoreFailed RestoreConditionType = "Failed"
+)
+
+// RestoreCondition describes the observed state of a Restore at a certain point.
+type RestoreCondition struct {
```
Comment (LinuxGit) [inline pkg/apis/pingcap.com/v1alpha1/types.go]: ditto, no json struct field tags

File: pkg/apis/pingcap.com/v1alpha1/types.go [hunk 5]
```diff
@@ -258,3 +258,215 @@ type TiKVFailureStore struct {
...
+	Message string
+}
+
+// RestoreSpec contains the specification for a restore of a tidb cluster backup.
+type RestoreSpec struct {
+	// Cluster represents the tidb cluster to be restored.
+	Cluster string `json:"cluster"`
+	// Backup represents the backup object to be restored.
+	Backup string `json:"backup"`
+	// SecretName is the name of the secret which stores
```
Comment (LinuxGit) [inline pkg/apis/pingcap.com/v1alpha1/types.go]: `Cluster` and `Backup` are string type, what information does they store?

Comment (onlymellb) [reply pkg/apis/pingcap.com/v1alpha1/types.go]: Point to TidbCluster CR and Backup CR name

File: pkg/apis/pingcap.com/v1alpha1/types.go (hunk 2 above)
Comment (onlymellb) [reply pkg/apis/pingcap.com/v1alpha1/types.go]: I think called `BackupSchedule` is ok, the periodic backup of other projects is also named

File: pkg/apis/pingcap.com/v1alpha1/types.go [hunk 6]
```diff
@@ -260,3 +260,215 @@ type TiKVFailureStore struct {
...
+	// RestoreFailed means the Restore has failed.
+	RestoreFailed RestoreConditionType = "Failed"
+)
+
+// RestoreCondition describes the observed state of a Restore at a certain point.
+type RestoreCondition struct {
+	Type    RestoreConditionType   `json:"type"`
+	Status  corev1.ConditionStatus `json:"status"`
+	Reason  string                 `json:"reason"`
+	Message string                 `json:"message"`
```
Comment (xiaojingchen) [inline pkg/apis/pingcap.com/v1alpha1/types.go]: this field should add `omitempty` tag

Comment (onlymellb) [reply pkg/apis/pingcap.com/v1alpha1/types.go]: It doesn't matter, address it in the next PR

**Pull request merged**